meshio/abaqus/_abaqus.py

```diff
@@ -111,10 +113,20 @@ def read_buffer(f):
     cell_data = {}
     point_data = {}
 
+    parent_file_gen = None
+
     line = f.readline()
     while True:
         if not line:  # EOF
-            break
+            if parent_file_gen is not None:
+                f.close()
```

Comment:
This part is a bit mysterious. Why do we close the file? Then after that, `f.readline()`? Something is fishy...

Reply:
I'll admit it does get a bit fishy. Inside the keyword=INCLUDE condition I store the current file generator inside the "parent_file_gen" variable and replace the variable "f" with a generator of the external file (as shown below).
 https://github.com/nschloe/meshio/blob/5e8f401868691e9215b18350ca9e27e947867a6c/meshio/abaqus/_abaqus.py#L180-L190

What the change in the EOF eval does is that it checks if there is stored anything in the temp "parent_file_gen" variable. If it is, then it will close the current generator (which is the external file), and assigns back the original generator to the variable "f" and moves beyond the "include" line with a new readline() statement.

https://github.com/nschloe/meshio/blob/5e8f401868691e9215b18350ca9e27e947867a6c/meshio/abaqus/_abaqus.py#L120-L129

I agree it is not pretty replacing variables like this, but for now I cant see any real alternatives that won't need changing other parts of the read_buffer potentially breaking dependencies from other readers (which was the case during my last attempt). Anyways, let me know if you find this acceptable, or you have some suggestions for any alternative implementations that you might want me to explore. Otherwise I guess you can close this pull request for now.

Reply:
Nope, too messy. Why can't we do something like the following? (Actual question.)
```python
_read(filename):
   # ...
   elif keyword == "INCLUDE": 
        include_filename = # read line and get filename
        out = _read(include_filename)
        # merge `out` into the mesh
```

Reply:
A good question. 

That was actually my ambition in my previous pull request. But unfortunately I made it (perhaps) unecessary complex by trying to account for points being defined within the external file AND the current file (which in hindsight would probably only apply when the model is using a "parts and assembly" model definition and\or multiple parts, and I am unsure if this is supported by meshio?). 

See my previous attempt (https://github.com/Krande/meshio/blob/35c012bdecf56d5e045bf2b5ef775ac3d1e5473b/meshio/abaqus/_abaqus.py#L180-L197) (excerpts shown below) where I ended up having to add an initial decleration of "points" variable as an empty numpy array and concatenate any points found in the current and external file. In the process I also think I messed up the "point_ids_in" also somehow (but on closer inspection, that variable might not be used anywhere else?). 

```python
def read_buffer(f):
    # Initialize the optional data fields
    points = numpy.empty([0, 3], dtype=numpy.float64)
    ...
    while True:
        ...
        elif keyword == "INCLUDE":
            incl_ref = line.lower().split("input=")[-1].rstrip().replace("\\", "/")
            if Path(incl_ref).exists() is False:
                cd = Path(f.name).parents[0]
                incl_ref = Path(cd, incl_ref).absolute()
            with open(incl_ref, "r") as d:
                submsh = read_buffer(d)


            points = numpy.concatenate((points, submsh.points))
            cells += submsh.cells
            point_data.update(submsh.point_data)
            cell_data.update(submsh.cell_data)
            field_data.update(submsh.field_data)
            point_sets.update(submsh.point_sets)
            cell_sets.update(submsh.cell_sets)


            line = f.readline()
        else:
```

I guess I could make it work by assuming that the points are either declared in the current file or the external file by simply doing something like `points = submsh.points`. But I might have to study the internal workings of the Mesh class first as to not break any potential inter-dependencies of list indices between the various lists and dicts.

Let me know what you think and if you have some suggestions while I contemplate and study the Mesh class :)

Reply:
This is the way it should work, except that you don't `concat` in each step, but just
```python
points = []
for [...]:
    points.append(more_points)
points = numpy.concatenate(points)
```
Or even
```
meshes = []
for [...]:
    meshes.append(new_mesh)
mesh = merge(meshes)
```
The problem with the previous PR was not this approach, but the fact you read an entire file at once. This is not necessary.

Reply:
Thanks a lot for taking the time helping me out! I really appreciate it! Okay, so I've given it another go and I've created a merge function and have started combining the meshes found in the external files with the "main" mesh containers in the read_buffer. I still have to figure out the cell_sets and CellBlock relationship and how to safely merge cell_sets (that was a bit more complex than I anticipated). 

Hopefully I will find some time in the next couple of days to fix that. In the mean time I thought it would be a good idea to push what I have so far. Let me know If you have any comments!